### DISCUSSION TITLE
Pearson correlation without z-normalization

### DISCUSSION THREAD

**Discussion by tomweb153** [TDAmeritrade/stumpy#932] [Help: Other Questions]
Hello everyone, 

I would like to ask, if it's possible to compute the pearson correlation without z-normalization using stumpy?

I have read something about the pearson correlation here: https://github.com/TDAmeritrade/stumpy/discussions/484

However, due to many flat sequences z-normalization becomes counterproductive for me.

Thanks!

**Comment by seanlaw** [comment]
@tomweb153 Thank you for your question and welcome to the STUMPY community.

It isn't so clear to me what you mean when you say "Pearson correlation without z-normalization". At the end of the day, [Pearson correlation](https://en.m.wikipedia.org/wiki/Pearson_correlation_coefficient) has a very clear definition (and does not necessarily depend on z-norm). In general, if your time series contains many constant subsequences then, by definition, the Pearson correlation will be 1.0 by simply following its definition above. Now, if you know that you are disinterested in those constant subsequences then you can change those constant sections to `np.nan` and then they will NEVER become a 1-nearest-neighbor to any subsequence.

Hopefully that helps. Otherwise, please feel free to provide a small working code example to demonstrate your point.

**Reply by tomweb153** [reply]
Hey @seanlaw and thank you.

I'd like to know if it's possible to get the values of the calculations of the Pearson Correlation and not the Euclidean distance that is based on z-normalization or not. I have tried searching for a line in your stumpy implementations where one could possible obtain the Pearson correlation (when not z-normalizing the subsequences -> flat) and return it, just like any stumpy function returns the mp.

Normally every stumpy function returns the Euclidean distance as it's the main similarity measure of the whole matrix profile idea, but I was wondering if using the Pearson correlation as a similarity measure is possible. And as the stumpy implementation is very fast I'd like to use the basis of the stumpy libraries.

**Reply by seanlaw** [reply]
@tomweb153 Consistent with what was demonstrated in #484, this is what you would do if you want to compute the Pearson correlation instead of z-norm Euclidean distance:

```
import stumpy
import numpy as np

T = np.random.rand(10_000)
m = 50

mp = stumpy.stump(T, m)
pearson = 1.0 - (np.square(mp[:, 0]) / (2.0 * m))
```

Is that what you are looking for? Since the conversion from z-norm Euclidean distance to Pearson correlation is so direct, there is no existing API function that will give you the Pearson correlation directly. Instead, you must convert the z-norm distance to Pearson correlation yourself. This should still be super fast/instantaneous once you've computed the z-norm matrix profile. Again, to go from z-norm Euclidean distance to Pearson correlation is only one additional line of code:

```
pearson = 1.0 - (np.square(mp[:, 0]) / (2.0 * m))
```

Note, I'm not sure if it is possible to convert non-normalized Euclidean distance (i.e., without z-norm or `normalize=False`) to Pearson correlation as there may not be a 1:1 correspondence. @NimaSarajpoor any thoughts on this point?

**Reply by tomweb153** [reply]
@seanlaw thank you for your response.

You are absolutely right with the instant transformation of z-normalized euclidean distance to Pearson Correlation. However, due to the type of my data (many flat regions, noisy subsequences) z-normalization of subsequences is not feasabile (normalize must be set to False). 

 See [Implications of z normalization on the Matrix Profile](https://link.springer.com/chapter/10.1007/978-3-030-40014-9_5#:~:text=Internally%2C%20the%20Matrix%20Profile%20utilizes,visual%20similarity%20of%20these%20subsequences). I have tried to implement the results of this paper, but the outcome wasnt satisfactory.

As I am in need of a fixed feature range (e.g. Pearson correlation is bounded to -1...1) but cannot utilize z-normalization on subsequences for easy transformation, I was asking, if there was any way one could bound the non-z-normalized Euclidean distance to any range.

**Reply by seanlaw** [reply]
@tomweb153 Are you able to provide a small concrete example (with code and data) so that we can better understand the problem? You should be able to drop a csv file into the comment box so that it gets uploaded and shared here.

**Reply by seanlaw** [reply]
> if there was any way one could bound the non-z-normalized Euclidean distance to any range.

<s>I think that the short answer is, "no". Unless you know what the absolute maximum allowable distance is, it is impossible to arbitrarily select a range. </s>

I vaguely recall reading somewhere something like, "for a given window size, `m`, the maximum window z-norm Euclidean distance is `sqrt(2.0*m)` but I can't recall which paper that came from. You can see some of this documented [here](https://github.com/TDAmeritrade/stumpy/blob/main/docs/Pearson.ipynb).

As a semi-related side note, say you had two identical subsequences that are of length `m=50` and so their Pearson correlation would be `1.0`. And, independently, you have two identical subsequences that are of length `m=5000` and so their Pearson correlation would also be `1.0`. There is no obvious or straightforward way to compare the subsequences of different lengths. I would almost argue that the subsequences with `m=5000` are, intuitively, "more" correlated as it is probabilistic less likely for two subsequences of that length to be identical. The point here is that you cannot force a range on your distances AND I would argue that Pearson correlation is even harder to interpret beyond the extreme values of `-1, 0, and +1`. What does a correlation of `0.5` mean when comparing subsequences?! Are the subsequences similar/dissimilar? I have no idea. This is something that you will need to decide on your own.

**Reply by tomweb153** [reply]
@seanlaw Thank you for your response. 

I tried to provide a minimal example. I have selected data from 6 sensors from 2 different systems. Measurements are taken every 5 minutes and the data spans over 1 week. Subsequence length is set to 12 (==1hour of data).

The code I provided calculates the (left) matrix profile for every sensor independently. As you can see the data is mostly steady with some outliers/anomalies present in data2.csv. I wish to detect these outliers/anomalies with the matrix profile (I provided a very easy case with some global point outliers).
When tweaking the `normalize` flag to `False` you can see high distances for the outliers/anomalies (e.g. around index 700 for arr2). However one cannot see any high mp-values when `normalize` is set to `True`.  For the sake of completeness I also added some data without any outliers/anomalies (see data1.csv).

I would love to set normalization to true, but I don't seem to be able to get my data into the right shape. I was hoping I could use a bounded output (e.g. range -1...1) so a threshold estimation could be done. The threshold is supposed to determine whether a anomaly may have manifested itself on the data or not.

Unfortunately I cannot paste my Jupyter-Notebook as a file, so I have to write the code here. You should be able to run this code in a Jupyter-Notebook with Python 3.10.0 installed.

    from stumpy import stumpi
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    
    df1 = pd.read_csv('data1.csv', delimiter=',', index_col='timestamp')
    df2 = pd.read_csv('data2.csv', delimiter=',', index_col='timestamp')
    
    arr1 = df1.to_numpy()
    arr2 = df2.to_numpy()
    
    fig, axs = plt.subplots(6, sharex=True, figsize=(30, 15))
    
    # change arr for different plot
    for i in range(6):
        axs[i].plot(arr2[:,i])
        plt.show()
        
    # calc lmp for each dimension independently
    # change arr1 to arr2 if needed
    dimensions = arr1.shape[1]
    all_matrix_profiles = []
    for dim in range(dimensions):
            
        n_time_series = arr1[:,dim]
        # change normalize to see "problems"
        n_matrix_profile = stumpi(n_time_series, m=12, normalize=False);
        n_lmp = n_matrix_profile.left_P_.astype(np.float64)
        
        # exclude warmup-phase 
        # -> exclude inf and -1 values
        n_lmp = n_lmp[12:]
    
        all_matrix_profiles.append(n_lmp)
        
        # transpose to get dimensions as columns and rows as index
    all_matrix_profiles = np.asarray(all_matrix_profiles).T
    
    # plot lmps
    fig, axs = plt.subplots(6, sharex=True, figsize=(30, 15))
    
    # change arr for different plot
    for i in range(6):
        axs[i].plot(all_matrix_profiles[:,i])
        plt.show()

[data2.csv](https://github.com/TDAmeritrade/stumpy/files/13192425/data2.csv)
[data1.csv](https://github.com/TDAmeritrade/stumpy/files/13192427/data1.csv)

**Reply by seanlaw** [reply]
So, one thing to understand is that matrix profiles with z-norm are good for finding subsequences with similar shape (regardless of their overall magnitude). However, they are horrible for finding anomalous (single point) spikes in your data. As you've identified, setting `normalize=False` is the only reasonable way to identify those spikes. 

> I was asking, if there was any way one could bound the non-z-normalized Euclidean distance to any range.

No, unfortunately there is no way to bound the non-normalized Euclidean distance .

**Reply by NimaSarajpoor** [reply]
@tomweb153 
Since you probably got most of your answers, I am going to just add two short notes.

> When tweaking the normalize flag to False you can see high distances for the outliers/anomalies (e.g. around index 700 for arr2). However one cannot see any high mp-values when normalize is set to True. For the sake of completeness I also added some data without any outliers/anomalies (see data1.csv).

I think you first need to understand what `normalize=False` means, and how the distance calculation works (e.g. see [this video](https://www.youtube.com/watch?v=T9_z7EpA8QM)). 

Let's say your time series is `T`, and length of subsequence is `m`. Also, let's say the minimum value in `T` is `MIN`, and the maximum value is `MAX`. It can be shown that the distance between two subsequences cannot be outside of the range $[0, (MAX-MIN)\sqrt{m}]$ (when `normalize` is `False`). 

When `normalize=True`, the distance between two subsequences is the distance between their z-normalized version (i.e. EACH of the two subsequences is Z-NORMALIZED first, and **_THEN_** their distance is being calculated). It can be shown that this distance is always in the range $[0, 2\sqrt{m}]$.  For each DISTANCE value (dissimilarity), you can find its CORRESPONDING SIMILARITY SCORE, which is just the "correlation" between the two subsequence. See [this doc](https://github.com/TDAmeritrade/stumpy/blob/main/docs/Pearson.ipynb) for further info.

Note: What is the lesson here? Sometimes it might be good to know, at some level, the underlying concept / logic. This can help you better understand the result. Otherwise, you will be forced to justify! 

> I was asking, if there was any way one could bound the non-z-normalized Euclidean distance to any range.

I do not know the "why" behind this. Maybe you have some reasons, and I am not sure if I understand your message correctly or not. So, I am going to give answer from my perspective based on my own understanding. I assume you just want to change the upper-bound $(MAX-MIN)\sqrt{m}$ (provided earlier above) to something else WHILE keeping the relative distance between subsequences untouched.  In this case, you can consider the factor $\frac{1}{(MAX-MIN)\sqrt{m}}$ and scale your whole time series by that, or maybe the general scaler $\frac{A}{B}$. I will leave it to you to think and see how that can affect that distances, and the range (and you, the user, should think if that makes sense for your problem or not).

**Reply by seanlaw** [reply]
> Let's say your time series is T, and length of subsequence is m. Also, let's say the minimum value in T is MIN, and the maximum value is MAX. It can be shown that the distance between two subsequences cannot be outside of the range 
 (when normalize is False).

Note that while this may be true for a static time series, this assumption will break if you have a new data point that is larger than `MAX` and/or smaller than `MIN` (and `normalize=False`). The point is that when `normalize=False`, this implies that new/additional data points in your time series is unbounded. This is similar to the problem of normalization in machine learning problems where you may cap your min/max range according to your training data in order to limit the range between `0` and `1` but your test data may contain values that are much larger than your `MAX` and therefore it will be normalized to something larger than `1` (or smaller than `MIN` and therefore will be normalized to something smaller than `0`).  Accordingly, you will run into the same issue if you use one time series to decide your `MIN`/`MAX` and then you find that a second time series has values outside of that range (i.e., either below the `MIN` OR above the `MAX`) then you'll get into trouble (again for `normalize=False`). So please proceed with caution.

**Reply by tomweb153** [reply]
@seanlaw @NimaSarajpoor thank you both for your insights.

I will try to take the things  you have said into account!
